synapse/appservice/api.py

```diff
@@ -493,11 +497,18 @@ async def claim_client_keys(
         # or if some are still missing.
         #
         # TODO This places a lot of faith in the response shape being correct.
-        missing = [
-            (user_id, device, algorithm)
-            for user_id, device, algorithm in query
-            if algorithm not in response.get(user_id, {}).get(device, [])
-        ]
+        missing = []
+        for user_id, device, algorithm, count in query:
+            # The number of keys responded for this algorithm.
+            response_count = sum(
```

Comment:
If Python changed this then loads of existing source code would break; I don't think they'll ever do this.

```
2023-04-27 13:45:19 ✔  $ python2.7
Python 2.7.18 (default, Dec 19 2022, 00:00:00) 
[GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> sum([True, False, True, False])
2
>>> 

2023-04-27 13:50:51 ✔  $ python3
Python 3.11.3 (main, Apr  5 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> sum([True, False, True, False])
2
```

Reply:
https://github.com/python/cpython/blob/main/Python/bltinmodule.c#L2508-L2517 suggests that the C implementation explicitly interprets bools as PyLong objects. In a fallback case it [defers to the `+` operator](https://github.com/python/cpython/blob/main/Python/bltinmodule.c#L2630) which also exhibits this behaviour: 

```python
>>> True + False
1
```

Reply:
Hum I thought it was counting all the items while it is counting the `True`. 
I am a bit stupid, but it also shows that the behavior is really not obvious.

Can we do `len(filter(None, ...))` instead ?

Reply:
> It may work in practice but it may break,

It is counting the number of `True` elements, yes. Booleans are integers and there has been no movement to change this in Python. I'm not concerned about treating them this way.

> Can we do `len(filter(None, ...))` instead ?

You need to make it a `list` to use `len(...)`; if we were doing that we could use `count(...)` instead like:

```python
list(
    key_id.startswith(f"{algorithm}:") for key_id in response.get(user_id, {}).get(device, {})
).count(True)
```

But I don't really see that as much simpler.

Another option would be to explictly return 1 or 0 from it, but again I find this more confusing:

```
sum(
    1 if key_id.startswith(f"{algorithm}:") else 0 for key_id in response.get(user_id, {}).get(device, {})
)
```

There's a comment above this statement which says:

> The number of keys responded for this algorithm.

Can I make this comment clearer so that the reader doesn't have to necessarily understand exactly how this code works?

Reply:
I found the `list().count(True)` a lot clearer TBH. It's pretty clear what it is doing then, and match the comment a lot better I found.
Or we could add a comment about the fact that summing boolean counts the `True`s. I don't know if it's common to assume in Python, but it's far from obvious to me. If we assume that the reader doesn't really want to understand this code however, current state is good enough.